src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/HtmlCSharp/CompletionHandler.cs

```diff
@@ -215,6 +229,63 @@ private bool IsSimpleImplicitExpression(LSPDocumentSnapshot documentSnapshot, Co
             return false;
         }
 
+        // We should remove Razor design time helpers from C#'s completion list. If the current identifier being targeted does not start with a double
+        // underscore, we trim out all items starting with "__" from the completion list. If the current identifier does start with a double underscore
+        // (e.g. "__ab[||]"), we only trim out common design time helpers from the completion list.
+        private SumType<CompletionItem[], CompletionList>? RemoveDesignTimeItems(
+            SumType<CompletionItem[], CompletionList> completionResult,
+            LSPDocumentSnapshot documentSnapshot,
+            TextExtent? wordExtent,
+            LanguageServerKind serverKind)
+        {
+            if (serverKind != LanguageServerKind.CSharp)
+            {
+                return completionResult;
+            }
+
+            var result = completionResult.Match<SumType<CompletionItem[], CompletionList>?>(
+                items => RemoveDesignTimeItems(items, documentSnapshot, wordExtent),
+                list => RemoveDesignTimeItems(list.Items, documentSnapshot, wordExtent));
+
+            return result;
+
+            static CompletionItem[] RemoveDesignTimeItems(CompletionItem[] items, LSPDocumentSnapshot documentSnapshot, TextExtent? wordExtent)
+            {
+                var designTimeHelpersCompletionItems = GenerateCompletionItems(DesignTimeHelpers);
+                if (wordExtent.HasValue)
+                {
+                    var wordCharacterStartIndex = wordExtent.Value.Span.Start.Position;
+                    if (wordCharacterStartIndex + 1 < documentSnapshot.Snapshot.Length)
+                    {
+                        var snapshot = documentSnapshot.Snapshot;
+
+                        // If the current identifier starts with "__", only trim out common design time helpers from the list.
+                        if (snapshot[wordCharacterStartIndex] == '_' && snapshot[wordCharacterStartIndex + 1] == '_')
+                        {
+                            return items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
+                        }
+                    }
+                }
+
+                // In all other cases, trim out both common design time helpers and all completion items starting with "__".
+                var updatedItems = items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
```

Comment:
```suggestion
                var filteredItems = items.Except(designTimeHelpersCompletionItems, CompletionItemComparer.Instance).ToArray();
```

Reply:
No matter what you do in this flow we're always going to filter out the common design time helpers so it may be valuable and more readable to do that at the top and then simplify the flow to something like:
```C#
var filteredItems = items.Except(razorDesignTimeHelpers, CompletionItemComparer.Instance).ToArray();
if (SuppressAllDesignTimeItems(documentSnapshot.Snapshot, wordExtent))
{
    var designTimeSuppressedItems = filteredItems.Where(item => item.InsertText != null && !item.InsertText.StartsWith("__", StringComparison.Ordinal)).ToArray();
    return designTimeSupprssedItems;
}

return filteredItems;
```